hydrogym/ts.py

```diff
@@ -143,18 +157,65 @@ def initialize_operators(self):
             proj_prob, solver_parameters={"ksp_type": "cg", "pc_type": "sor"}
         )
 
+    def set_damping(self, k_new):
+        if not isinstance(k_new, list):
+            k_new = [k_new]
+        self.controller_damping_coeff = k_new
+        return
+
+    def get_inertia(self):
+        return self.inertia
+
+    def get_state(self):
+        return self.state
+
     def update_controls(self, control):
-        """Add a damping factor to the controller response
+        if self.control_method == "indirect":
+            if self.flow.name == "Cylinder":
+                """
+                Update BCS of cylinder to new angular velocity using implicit euler solver according to diff eqn:
+
+                omega_t[i+1] = omega_t[i] + (d_omega/dt)_t[i+1] * dt
+                            = omega_t[i] + (control_t[i+1] - k_damping*omega_t[i+1])/I_cm * dt
+
+                omega_t[i+1] can be solved for directly in order to avoid using a costly root solver
+                """
+                next_state = []
+                for (state, ctrl, I_cm, k_damping) in zip(
+                    self.state, control, self.I_cm, self.k_damping
+                ):
+                    next_state.append(
+                        (state + ctrl * self.dt / I_cm)
+                        / (1 + k_damping * self.dt / I_cm)
+                    )
+                self.state = next_state
+
+            elif self.flow.name == "Pinball":
+                raise Exception(
+                    "Indirect control of Pinball environment is not yet supported"
+                )
+            elif self.flow.name == "Cavity":
+                raise Exception(
+                    "Indirect control of Cavity environment is not yet supported"
+                )
+            elif self.flow.name == "Step":
+                raise Exception(
+                    "Indirect control of Step environment is not yet supported"
+                )
+            else:
+                raise Exception(
+                    "Unknown Environment Selected (indirect control is not yet supported)"
+                )
```

Comment:
Again, I think we should avoid having the timestepper depend on any details of the flow configuration.  For instance, if someone wanted to add their own flow configuration, they should just be able to implement a new subclass of `FlowConfig` without modifying the source of `ts.py` or anything else in the `hydrogym` package.

But it looks to me like this should be fine, assuming that these configurations provide methods like `get_damping` and `get_inertia`.  This looks like it should work with any number of inputs, which is the main thing.  I'd just maybe suggest a minor renaming to make it more general, since the equation for F=m*a also basically works for tau=I*\dot{omega} (I think this would also fix the comments above)

```
for (state, ctrl, m, k) in zip(
                    self.state, control, self.inertia, self.controller_damping_coef
                ):
    state = (state + ctrl * self.dt / m) / (1 + k* self.dt / m)
```

Reply:
And maybe `get_damping` and `get_inertia` could raise a `NotImplementedError` similar to what you have here by default in `FlowConfig` until `Cavity`, `Pinball`, etc implement these.